packages/code-editor/src/CodeEditor/hooks/useLazyModules.ts

```diff
@@ -0,0 +1,132 @@
+ *
+ * if (!isLoading && modules.utils) {
+ *   modules.utils.helper();
+ * }
+ * ```
+ */
+export const useLazyModules = <T extends object>(
+  loaders: LoadersMap<Partial<T>>,
+) => {
+  const [modules, setModules] = useState<Partial<T>>({});
+  const [isLoading, setIsLoading] = useState<boolean>(false);
+
+  useEffect(() => {
+    /*
+     * Get keys from loaders object and cast them to the generic type
+     * This type assertion is necessary because Object.keys returns string[]
+     */
+    const neededKeys = Object.keys(loaders) as Array<keyof T>;
+    const availableKeys = Object.keys(modules) as Array<keyof T>;
+
+    /*
+     * Determine which modules need to be loaded to prevent duplicate loads
+     */
+    const keysToLoad = neededKeys.filter(key => !availableKeys.includes(key));
+
+    /*
+     * Early return if there's nothing to load
+     */
+    if (keysToLoad.length === 0) {
+      setIsLoading(false);
+      return;
+    }
+
+    setIsLoading(true);
+
+    /*
+     * Create an array of promises for each module that needs to be loaded
+     * The non-null assertion (!) is used because TypeScript can't guarantee
+     * the key exists in loaders, even though we filtered for keys in loaders
+     */
+    const modulePromises = keysToLoad.map(async key => {
+      try {
+        const module = await loaders[key]!();
+        return { key, module };
+      } catch (err: any) {
+        console.error(`Error loading module "${String(key)}":`, err);
+      }
+    });
+
+    (async () => {
+      try {
+        const modules = await Promise.all(modulePromises);
+
+        /*
+         * Filter out any failed loads (which returned undefined)
+         * The expression below uses a TypeScript type predicate with the syntax:
+         * (parameter): parameterType is SpecificType => boolean
+         * This tells the compiler that when the function returns true,
+         * the parameter should be treated as the specified type.
+         * Without this predicate, TypeScript would not know that filtered modules
+         * are guaranteed to have the key and module properties.
+         */
+        const validModules = modules.filter(
+          (module): module is { key: keyof T; module: any } => !!module,
+        );
+
+        /**
+         * Convert array of key-module pairs to an object
+         * The type assertion for the initial value ({} as Partial<T>) is necessary
+         * because an empty object literal {} has type {}, not Partial<T>.
+         * This ensures the accumulator (acc) has the correct type from the start.
+         */
+        const newModules = validModules.reduce((acc, { key, module }) => {
+          acc[key] = module;
+          return acc;
+        }, {} as Partial<T>);
+
+        setModules(prev => ({ ...prev, ...newModules }));
+      } catch (err: any) {
+        console.error(`Failed to load module:`, err);
+      } finally {
+        setIsLoading(false);
+      }
+    })();
+  }, [loaders, modules]);
```

Comment:
This effect makes a call to `setModules` which will trigger it to run again. The early return will prevent the infinite loop, but this is a bit of a code-smell.

I'd recommend replacing the `modules` state with a reducer, and moving the bulk of this pre-computation logic into that reducer. Then all the effect needs to do is dispatch an action when `loaders` changes